File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    public int solution(int[][] maps) {
        int n = maps.length;
        int m = maps[0].length;
        
        int [][] visited = new int[n][m];
        for (int[] row : visited){
            Arrays.fill(row,-1);
        }
        
        int[] dx = {-1,1,0,0};
        int[] dy = {0,0,-1,1};
        
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{0,0});
        visited[0][0]=1;
        
        while (!queue.isEmpty()){
            int[] cur = queue.poll();
            int x = cur[0];
            int y = cur[1];
            
            for (int dir=0; dir<4; dir++){
                int nx = x+dx[dir];
                int ny = y+dy[dir];
                
                if (nx<0 || nx>=n || ny<0 || ny>=m){
                    continue;
                }
                if (maps[nx][ny]==0){
                    continue;
                }
                if (visited[nx][ny]!= -1){
                    continue;
                }
                visited[nx][ny]=visited[x][y]+1;
                queue.offer(new int[]{nx,ny});
            }
        }
        return visited[n-1][m-1];
    }
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    n=len(maps)
    m=len(maps[0])
    
    visited=[[False]*m for _ in range(n)]
    queue=deque()
    queue.append((0,0,1))
    visited[0][0]=True
    
    dx=[-1,1,0,0]
    dy=[0,0,-1,1]
    
    while queue:
        x,y,dist=queue.popleft()
        if x==n-1 and y==m-1:
            return dist
        
        for i in range(4):
            nx,ny=x+dx[i], y+dy[i]
            
            if 0<= nx <n and 0<=ny<m:
                if maps[nx][ny]==1 and not visited[nx][ny]:
                    visited[nx][ny]=True
                    queue.append((nx,ny,dist+1))

    return -1